File: fsharp/pascals-triangle/PascalsTriangle.fs
module PascalsTriangle

    let private cons (head:'a) (tail:'a list) = head::tail

    let rec private scanner (differenceList:int list list -> int list list) (row:int list) (n:int):int list list =
        let newRow = List.map2(fun element1 element2 -> element1 + element2) (0::row) (0::row |> List.rev)
        match n with
        | 1 -> (differenceList << cons newRow) []
        | _ -> scanner (differenceList << cons newRow) newRow (n - 1)

    let rec private pascalsTriangle (k:int):int list list =
        match k with
        | 0 -> []
        | 1 -> [[1]]
        | _ -> scanner (id << cons [1]) ([1]) (k - 1)

    let rows (numberOfRows:int):int list list option =
        match numberOfRows > -1 with
        | true ->
            numberOfRows
            |> pascalsTriangle
            |> Some
        | false -> None

File: solutions/scala/pascals-triangle/1/src/main/scala/PascalsTriangle.scala
object PascalsTriangle
{

    private def scanner (stack:List[List[Int]]) (row:List[Int]) (n:Int):List[List[Int]] =
    {
        def extended:List[Int] = 0::row
        def extendedReverse:List[Int] = extended.reverse
        def newRow:List[Int] = extended.zip(extendedReverse).map(elements => elements._1 + elements._2)
        n match
        {
            case 1 => (newRow::stack).reverse
            case _ => scanner (newRow::stack) (newRow) (n - 1)
        }
    }

    private def pascalsTriangle (k:Int):List[List[Int]] =
    {
        k match
        {
            case 0 => Nil
            case 1 => List(List(1))
            case _ => scanner (List(List(1))) (List(1)) (k - 1)
        }
    }

    def rows (numberOfRows:Int):List[List[Int]] =
    {
        numberOfRows > -1 match
        {
            case true => pascalsTriangle(numberOfRows)
            case false => Nil
        }
    }
}